aurora/core.py

```diff
@@ -1189,73 +1339,341 @@ def check_conservation(self, plot=True, axs=None, plot_resolutions=False):
+        Returns
+        -------
+        reservoirs : dict
+            Dictionary containing density of particles in each reservoir time-averaged over cycles
+        axs : matplotlib.Axes instances, only returned if plot=True
+            Array-like structure containing two matplotlib.Axes instances, (ax1,ax2).
+            See optional input argument.
+        """
+        (
+            nz,
+            N_wall,
+            N_div,
+            N_pump,
+            N_out,
+            N_ret,
+            N_tsu,
+            N_dsu,
+            N_dsul,
+            rcld_rate,
+            rclp_rate,
+            rclw_rate,
+        ) = self.res
+        nz = nz.transpose(2, 1, 0)  # time,nZ,space
+
+        # factor to account for cylindrical geometry:
+        circ = 2 * np.pi * self.Raxis_cm  # cm
+
+        # collect all the relevant quantities for particle conservation
+        reservoirs = {}
+
+        # calculate total impurity density (summed over charge states)
+        total_impurity_density = np.nansum(nz, axis=1)  # time, space
+
+        # Compute total number of particles for particle conservation checks:
+        all_particles = grids_utils.vol_int(
+            total_impurity_density, self.rvol_grid, self.pro_grid, self.Raxis_cm,
+            rvol_max = None
+        )
+
+        time_average, reservoirs["total"] = plot_tools.time_average_reservoirs(self.time_out,all_particles + (N_wall + N_div + N_pump + N_out + N_ret) * circ,time_start,time_end,interval)
+
+        # main fluxes
+        time_average, reservoirs["source"] = plot_tools.time_average_reservoirs(self.time_out,self.total_source*circ + reservoirs["total"][0],time_start,time_end,interval)
+        time_average, reservoirs["wall_source"] = plot_tools.time_average_reservoirs(self.time_out,rclw_rate * circ,time_start,time_end,interval)
+        time_average, reservoirs["divertor_recycling"] = plot_tools.time_average_reservoirs(self.time_out,rcld_rate * circ + rclp_rate * circ,time_start,time_end,interval)       
+        time_average, reservoirs["plasma_removal_rate"] = plot_tools.time_average_reservoirs(self.time_out,- N_dsu * circ - N_tsu * circ - N_dsul * circ,time_start,time_end,interval)    
+        reservoirs["net_plasma_flow"] = reservoirs["source"] + reservoirs["wall_source"] + reservoirs["divertor_recycling"] + reservoirs["plasma_removal_rate"]
+
+        # integrated source over time
+        reservoirs["integ_source"] = cumtrapz(reservoirs["source"], time_average, initial=0)
+
+        # main plasma content
+        if self.namelist["phys_volumes"]: 
+            ones = np.ones_like(total_impurity_density)
+            den = grids_utils.vol_int(ones[:,:],self.rvol_grid,self.pro_grid,self.Raxis_cm,rvol_max=self.rvol_lcfs)
+            time_average, reservoirs["particle_density_in_plasma"] = plot_tools.time_average_reservoirs(self.time_out,all_particles/den,time_start,time_end,interval)      
```

Comment:
@antonellozito any reason why this `time_average_reservoirs` function should be in the `plot_tools` script?